arangod/GeneralServer/H2CommTask.cpp

```diff
@@ -0,0 +1,654 @@
+  auto authToken = this->checkAuthHeader(*req);
+
+  // We want to separate superuser token traffic:
+  if (req->authenticated() && req->user().empty()) {
+    RequestStatistics::SET_SUPERUSER(this->statistics(1UL));
+  }
+
+  // first check whether we allow the request to continue
+  CommTask::Flow cont = this->prepareExecution(authToken, *req);
+  if (cont != CommTask::Flow::Continue) {
+    return;  // prepareExecution sends the error message
+  }
+
+  // unzip / deflate
+  if (!this->handleContentEncoding(*req)) {
+    this->addErrorResponse(rest::ResponseCode::BAD, req->contentTypeResponse(),
+                           1, TRI_ERROR_BAD_PARAMETER, "decoding error");
+    return;
+  }
+
+  // create a handler and execute
+  auto resp = std::make_unique<HttpResponse>(rest::ResponseCode::SERVER_ERROR,
+                                             req->messageId(), nullptr);
+  resp->setContentType(req->contentTypeResponse());
+  resp->setContentTypeRequested(req->contentTypeResponse());
+
+  this->executeRequest(std::move(req), std::move(resp));
+}
+
+namespace {
+bool expectResponseBody(int status_code) {
+  return status_code == 101 ||
+         (status_code / 100 != 1 && status_code != 304 && status_code != 204);
+}
+}  // namespace
+
+template <SocketType T>
+void H2CommTask<T>::sendResponse(std::unique_ptr<GeneralResponse> baseRes,
+                                 RequestStatistics* stat) {
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+  HttpResponse& response = dynamic_cast<HttpResponse&>(*baseRes);
+#else
+  HttpResponse& response = static_cast<HttpResponse&>(*baseRes);
+#endif
+
+  const int32_t sid = static_cast<int32_t>(response.messageId());
+  Stream* strm = findStream(sid);
+  if (strm == nullptr) {
+    LOG_TOPIC("9912b", ERR, Logger::REQUESTS)
+        << "response with message id " << response.messageId() << "has not H2 stream";
+    TRI_ASSERT(false);
+    return;
+  }
+
+  this->finishExecution(*baseRes, strm->origin);
+
+  if (!ServerState::instance()->isDBServer()) {
+    // DB server is not user-facing, and does not need to set this header
+    // use "IfNotSet" to not overwrite an existing response header
+    response.setHeaderNCIfNotSet(StaticStrings::XContentTypeOptions, StaticStrings::NoSniff);
+  }
+
+  // append write buffer and statistics
+  double const totalTime = RequestStatistics::ELAPSED_SINCE_READ_START(stat);
+
+  // and give some request information
+  LOG_TOPIC("924cc", DEBUG, Logger::REQUESTS)
+      << "\"h2-request-end\",\"" << (void*)this << "\",\""
+      << this->_connectionInfo.clientAddress
+      << "\",\""
+      //      << GeneralRequest::translateMethod(::llhttpToRequestType(&_parser))
+      << "\",\"" << static_cast<int>(response.responseCode()) << "\","
+      << Logger::FIXED(totalTime, 6);
+
+  strm->response.reset(static_cast<HttpResponse*>(baseRes.release()));
+
+  int32_t mid = static_cast<int32_t>(strm->response->messageId());
+  while (!_waitingResponses.push(mid)) {
```

Comment:
do we need to add some abort criteria here, such as if the server is going into shutdown?